google/cloud/storage/transfer_manager.py

```diff
@@ -0,0 +1,188 @@
+                results.append(exp)
+                continue
+        results.append(future.result())
+    return results
+
+
+def download_many(
+    blob_file_pairs,
+    download_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    if download_kwargs is None:
+        download_kwargs = {}
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+        for blob, path_or_file in blob_file_pairs:
+            method = (
+                blob.download_to_filename
+                if isinstance(path_or_file, str)
+                else blob.download_to_file
+            )
+            futures.append(executor.submit(method, path_or_file, **download_kwargs))
+    results = []
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        if not raise_exception:
+            exp = future.exception()
+            if exp:
+                results.append(exp)
+                continue
+        results.append(future.result())
+    return results
+
+
+def download_chunks_concurrently_to_file(
+    blob,
+    file_obj,
+    chunk_size=200 * 1024 * 1024,
+    max_workers=None,
+    download_kwargs=None,
+    deadline=None,
+):
+    if download_kwargs is None:
+        download_kwargs = {}
+    # We must know the size of the object, and the generation.
+    if not blob.size or not blob.generation:
+        blob.reload()
+
+    def download_range_via_tempfile(blob, start, end, download_kwargs):
+        tmp = tempfile.TemporaryFile()
+        blob.download_to_file(tmp, start=start, end=end, **download_kwargs)
+        return tmp
+
+    futures = []
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        cursor = 0
+        while cursor < blob.size:
+            start = cursor
+            cursor = min(cursor + chunk_size, blob.size)
+            futures.append(
+                executor.submit(
+                    download_range_via_tempfile,
+                    blob,
+                    start=start,
+                    end=cursor - 1,
+                    download_kwargs=download_kwargs,
+                )
+            )
+
+    # Wait until all futures are done and process them in order.
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        tmp = future.result()
+        tmp.seek(0)
+        file_obj.write(tmp.read())
+        tmp.close()
+
+
+def upload_many_from_filenames(
+    bucket,
+    filenames,
+    root,
+    prefix="",
+    skip_if_exists=False,
+    blob_constructor_kwargs=None,
+    upload_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    file_blob_pairs = []
+
+    for filename in filenames:
+        path = root + filename
+        blob_name = prefix + filename
```

Comment:
I think you will want to make sure filenames don't accidentally fall outside of root similar to https://github.com/advisories/GHSA-c28r-hw5m-5gv3;

You'll want to protect destination root so a filename doesn't attempt to write data outside of the expected location. Right now, this code IIRC fall to the same issue linked above.

@BenWhitehead had originally brought this up originally (thanks Ben).

Reply:
That's in the design doc under "Directory traversal" - of course, still open for debate. Let's take this discussion back to that doc if needed so it's all in one place.